### PR TITLE
New API /authenticators/custom to manage user defined local authenticators

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Issue: 
- https://github.com/wso2/product-is/issues/21216

With this PR, we add new API /authenticators/custom/ manage user defined local authenticators.

Service layer PR:
- https://github.com/wso2/carbon-identity-framework/pull/6172

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 1]
```diff
@@ -192,6 +204,122 @@ public ConnectedApps getConnectedAppsOfLocalAuthenticator(String authenticatorId
                     .ERROR_CODE_ERROR_RETRIEVING_IDP_CONNECTED_APPS, authenticatorId);
         }
     }
+    
+    public Authenticator addUserDefinedLocalAuthenticator(
+            UserDefinedLocalAuthenticatorCreation config) {
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:209]: bring to single line

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 2]
```diff
@@ -192,6 +204,122 @@ public ConnectedApps getConnectedAppsOfLocalAuthenticator(String authenticatorId
...
+        authenticator.setIsEnabled(config.isEnabled());
+        authenticator.setDefinedBy(Authenticator.DefinedByEnum.USER);
+        authenticator.setType(Authenticator.TypeEnum.LOCAL);
+        authenticator.setSelf(ContextLoader.buildURIForBody(String.format(V1_API_PATH_COMPONENT +
+                AUTHENTICATOR_PATH_COMPONENT + "/%s", config.getName())).toString());
+
+        return authenticator;
+    }
+
+    private UserDefinedLocalAuthenticatorConfig userDefinedLocalAuthenticatorForUpdate(
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:271]: rename to userDefinedLocalAuthenticatorForCreate

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 3]
```diff
@@ -192,6 +204,122 @@ public ConnectedApps getConnectedAppsOfLocalAuthenticator(String authenticatorId
...
+                            userDefinedLocalAuthenticatorForUpdate(config, existingAuthenticator),
+                            tenantDomain);
+            return buildAuthenticator(updatedConfig);
+        } catch (AuthenticatorMgtException e) {
+            throw handleAuthenticatorException(e, Constants.ErrorMessage.ERROR_CODE_ERROR_RETRIEVING_IDP_CONNECTED_APPS,
+                    authenticatorId);
+        }
+    }
+
+    private Authenticator buildAuthenticator(UserDefinedLocalAuthenticatorConfig config) {
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:257]: move to a util class

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.common/src/main/java/org/wso2/carbon/identity/api/server/authenticators/common/AuthenticatorsServiceHolder.java [hunk 4]
```diff
@@ -77,4 +79,24 @@ public void setIdentityProviderManager(IdentityProviderManager identityProviderM
 
         AuthenticatorsServiceHolder.getInstance().identityProviderManager = identityProviderManager;
     }
+
+    /**
+     * Get ApplicationAuthenticatorService osgi service.
+     *
+     * @return ApplicationAuthenticatorService
+     */
+    public ApplicationAuthenticatorService getApplicationCommonService() {
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.common/src/main/java/org/wso2/carbon/identity/api/server/authenticators/common/AuthenticatorsServiceHolder.java:88]: Why this method name is 'getApplicationCommonService' ? This is basically of retrieving "ApplicationAuthenticationService"

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml [hunk 5]
```diff
@@ -226,6 +374,92 @@ components:
...
+            - BEARER
+            - API_KEY
+            - BASIC
+          example: BASIC
+        properties:
+          type: object
+          additionalProperties: true
+          example:
+            username: "auth_username"
+            password: "auth_password"
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml:462]: Mark read only fields

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/impl/LocalAuthenticatorConfigBuilderFactory.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        } catch (NoSuchElementException | IllegalArgumentException e) {
+            Constants.ErrorMessage error = ERROR_CODE_INVALID_ENDPOINT_CONFIG;
+            throw new AuthenticatorMgtClientException(error.getCode(), error.getMessage(), e.getMessage());
+        }
+    }
+
+    private static AuthenticatorPropertyConstants.AuthenticationType resolveAuthenticationType(
+            LocalAuthenticatorConfig config) {
+
+        if (Arrays.asList(config.getTags()).contains("2FA")) {
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/impl/LocalAuthenticatorConfigBuilderFactory.java:136]: Make this constant

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/impl/LocalAuthenticatorConfigBuilderFactory.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        Authenticator authenticator = new Authenticator();
+        authenticator.setName(config.getName());
+        authenticator.setId(base64URLEncode(config.getName()));
+        authenticator.setDisplayName(config.getDisplayName());
+        authenticator.setIsEnabled(config.isEnabled());
+        authenticator.setDefinedBy(Authenticator.DefinedByEnum.USER);
+        authenticator.setType(Authenticator.TypeEnum.LOCAL);
+        authenticator.setTags(Arrays.asList(config.getTags()));
+        authenticator.setSelf(ContextLoader.buildURIForBody(String.format(V1_API_PATH_COMPONENT +
+                CONFIGS_AUTHENTICATOR_PATH_COMPONENT + "/%s", config.getName())).toString());
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/impl/LocalAuthenticatorConfigBuilderFactory.java:65]: Fix %s

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml [hunk 8]
```diff
@@ -118,7 +118,155 @@ paths:
...
+          description: Bad Request
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Error'
+        '401':
+          description: Unauthorized
+        '403':
+          description: Forbidden
+        '409':
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/resources/authenticators.yaml:151]: If this is not actually, remove this

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 9]
```diff
@@ -383,9 +463,9 @@ private List<String> getDistinctTags(IdentityProvider identityProvider) {
         if (fedAuthConfigs != null) {
             for (FederatedAuthenticatorConfig config : fedAuthConfigs) {
                 if (config.isEnabled()) {
-                    FederatedAuthenticatorConfig federatedAuthenticatorConfig =
-                            ApplicationAuthenticatorService.getInstance()
-                                    .getFederatedAuthenticatorByName(config.getName());
+                    FederatedAuthenticatorConfig federatedAuthenticatorConfig = AuthenticatorsServiceHolder
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:466]: get service move to private method.

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java [hunk 10]
```diff
@@ -908,6 +988,45 @@ private APIError handleIdPException(IdentityProviderManagementException e,
...
+    /**
+     * Handle IdentityProviderManagementException, extract error code, error description and status code to be sent
+     * in the response.
+     *
+     * @param e         IdentityProviderManagementException.
+     * @return APIError.
+     */
+    private APIError handleAuthenticatorException(AuthenticatorMgtException e) {
+
+        ErrorResponse errorResponse = new ErrorResponse.Builder().withCode(e.getErrorCode()).withMessage(e.getMessage())
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/core/ServerAuthenticatorManagementService.java:1000]: break at eacj `.`

File: components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/impl/LocalAuthenticatorConfigBuilderFactory.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        try {
+            UserDefinedAuthenticatorEndpointConfig.UserDefinedAuthenticatorEndpointConfigBuilder endpointConfigBuilder =
+                    new UserDefinedAuthenticatorEndpointConfig.UserDefinedAuthenticatorEndpointConfigBuilder();
+            endpointConfigBuilder.uri(endpointConfig.getUri());
+            endpointConfigBuilder.authenticationType(endpointConfig.getAuthentication().getType().toString());
+            endpointConfigBuilder.authenticationProperties(endpointConfig.getAuthentication().getProperties()
+                    .entrySet().stream().collect(Collectors.toMap(
+                            Map.Entry::getKey, entry -> entry.getValue().toString())));
+            return endpointConfigBuilder.build();
+        } catch (NoSuchElementException | IllegalArgumentException e) {
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/impl/LocalAuthenticatorConfigBuilderFactory.java:129]: Modify endpoint builder to check those exceptions and throw client exception

Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.api.server.authenticators/org.wso2.carbon.identity.api.server.authenticators.v1/src/main/java/org/wso2/carbon/identity/api/server/authenticators/v1/impl/LocalAuthenticatorConfigBuilderFactory.java:129]: The endpointConfigBuilder is common model for both federated and local authenticators, hence will not throw the exception from that model class.

**Review by malithie — APPROVED** [review]

**Pull request merged**